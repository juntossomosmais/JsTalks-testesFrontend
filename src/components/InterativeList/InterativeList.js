import React, { useState } from "react";
import { v4 as uuidv4 } from "uuid";
import { useForm } from "react-hook-form";

import * as Styled from "./InterativeList.styles";

const InterativeList = () => {
  const [listState, setListState] = useState([]);
  const { register, handleSubmit } = useForm();

  const onSubmit = (data) => {
    if (!data.item) return;
    setListState([...listState, data.item]);
  };

  const removeItem = (itemToRemove) => {
    const updatedList = listState.filter((item) => item !== itemToRemove);
    setListState(updatedList);
  };

  return (
    <>
      <Styled.ListWrapper>
        {listState.length > 0 ? (
          <ul>
            {listState.map((item) => {
              return (
                <li key={uuidv4()}>
                  {item} <button onClick={() => removeItem(item)}>X</button>
                </li>
              );
            })}
          </ul>
        ) : (
          <div>No Items :(</div>
        )}
      </Styled.ListWrapper>

      <Styled.FormWrapper onSubmit={handleSubmit(onSubmit)}>
        <input name="item" ref={register} />
        <button type="submit">Add item</button>
      </Styled.FormWrapper>
    </>
  );
};

export default InterativeList;
