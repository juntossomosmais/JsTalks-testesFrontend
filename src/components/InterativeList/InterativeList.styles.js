import styled from "styled-components";

export const ListWrapper = styled.div`
  display: flex;
  justify-content: center;
  margin-bottom: 30px;
  width: 100%;

  ul {
    margin: 0;
    padding: 0;
    list-style: none;
    width: 250px;

    li {
      display: flex;
      border-bottom: solid 1px #95a5a6;
      padding: 15px 0;
      justify-content: space-between;

      &:last-of-type {
        border: none;
      }
    }
  }
`;

export const FormWrapper = styled.form`
  display: flex;
  width: 100%;
  justify-content: center;

  input {
    padding: 5px;
    width: 200px;
    margin: 0 10px 0 0;
  }

  button {
    background: #2ecc71;
    border-radius: 4px;
    color: #fff;
    padding: 8px 12px;
    outline: none;
  }
`;
