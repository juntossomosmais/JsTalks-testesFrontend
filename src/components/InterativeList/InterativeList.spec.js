import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import InterativeList from "./InterativeList";

describe("<InterativeList />", () => {
  it("should display no items", () => {
    // when
    const { getByTestId, baseElement } = render(<InterativeList />);

    // then
    expect(getByTestId("no-items")).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot("no items");
  });

  it("should add a new item", async () => {
    // given
    const inputValue = "react test";

    // when
    const { getByTestId, rerender, getByText } = render(<InterativeList />);
    const input = getByTestId("add-item-input");
    const submitButton = getByTestId("add-item-sbmt");

    fireEvent.change(input, { target: { value: inputValue } });
    await waitFor(() => fireEvent.click(submitButton));

    rerender(<InterativeList />);

    const list = getByTestId("interative-list");

    // then
    expect(getByText(inputValue)).toBeInTheDocument();
    expect(list).toMatchSnapshot("item with the inputValue");
  });
});
