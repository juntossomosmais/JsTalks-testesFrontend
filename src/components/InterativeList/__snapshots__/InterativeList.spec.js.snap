// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InterativeList /> should add a new item: item with the inputValue 1`] = `
<ul
  data-testid="interative-list"
>
  <li>
    react test
  </li>
</ul>
`;

exports[`<InterativeList /> should display no items: no items 1`] = `
<body>
  <div>
    <div
      class="sc-AxjAm cqBkwT"
    >
      <div
        data-testid="no-items"
      >
        No Items :(
      </div>
    </div>
    <form
      class="sc-AxirZ jobyAr"
    >
      <input
        data-testid="add-item-input"
        name="item"
      />
      <button
        data-testid="add-item-sbmt"
        type="submit"
      >
        Add item
      </button>
    </form>
  </div>
</body>
`;
